### PR TITLE
Allow seeding MersenneTwisterEngine with arrays/ranges

### PR DESCRIPTION
**Pull request by n8sh**
Behavior is like the reference implementation's init_by_array,
not like Phobos std.random.MersenneTwisterEngine's seed(Range).

### CODE REVIEW

**Comment by 9il** [comment]
@n8sh Could you please rework it to
three generic methods:

The first one accepts `scope Slice!(Contiguous, [1], const(UInt)*)`;

the second one accepts an `scope const (UInt)[]`, applies `.sliced`, and passes it to the first one;

the third accepts an input range (no need to be forward), fills a local buffer from it and passes this buffer to the second one method.

**Comment by n8sh** [comment]
@9il I'm willing to make the change but there may be something I am misunderstanding.

> The first one accepts scope Slice!(Contiguous, [1], const(UInt)*);
> the second one accepts an scope const (UInt)[], applies .sliced, and passes it to the first one;

The code I have written that uses `scope Slice!(Contiguous, [1], const(UIntType)*)` looks like this:
```
        size_t i = n - 2;
        size_t j = 0;
        immutable key_length = slice.length;
        const(UIntType)[] field = slice.field; // <-- ???
        foreach_reverse(_; 0 .. (n > key_length ? n : key_length))
        {
            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f2))
                + cast(UIntType)(field[j] + j);
            static if (max != UIntType.max)
                data[i] &= max;
            if (--i > n)
            {
                data[$ - 1] = data[0];
                i = n - 2;
            }
            if (++j >= key_length)
                j = 0;
        }
```
Is it poor style to use `slice.field` to get `const(UIntType)[]` here? It seems odd to wrap `const(UIntType)[]` in a Slice struct then immediately unwrap it.

**Comment by 9il** [comment]
Just uses slice instead, it should just work. `slice.length` is zero cost.
```d
        import mir.utility: min;
        size_t i = n - 2;
        size_t j = 0;
        foreach_reverse(_; 0 .. min(slice.length, n))
        {
            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f2))
                + cast(UIntType)(slice[j] + j);
            static if (max != UIntType.max)
                data[i] &= max;
            if (--i > n)
            {
                data[$ - 1] = data[0];
                i = n - 2;
            }
            if (++j >= slice.length)
                j = 0;
        }
```

**Comment by 9il** [comment]
BTW, could you please slightly optimise the loop?
```d
        import mir.utility: min;
        size_t j = 0;
        immutable len = min(slice.length, n) - 1;
        foreach_reverse(i; 0 .. len) // do not handle tail
        {
            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f2))
                + cast(UIntType)(slice[j] + (len - 1 - i));
            static if (max != UIntType.max)
                data[i] &= max;
        }
       // handle tail conditions here
```

File: source/mir/random/engine/mersenne_twister.d [hunk 1]
```diff
@@ -90,6 +106,198 @@ struct MersenneTwisterEngine(UIntType, size_t w, size_t n, size_t m, size_t r,
...
+                j = 0;
+        }
+
+        foreach_reverse(_; 0 .. n-1)
+        {
+            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f3))
+                - cast(UIntType)(n - (i + 1)); /* non linear */
+            static if (max != UIntType.max)
+                data[i] &= max;
+            if (--i > n)
```
Comment (9il) [inline source/mir/random/engine/mersenne_twister.d]: the same here

File: source/mir/random/engine/mersenne_twister.d [hunk 2]
```diff
@@ -90,6 +106,198 @@ struct MersenneTwisterEngine(UIntType, size_t w, size_t n, size_t m, size_t r,
...
+    Constructs a MersenneTwisterEngine object.
+
+    This method copies the values from the range into a temporary buffer.
+    If you can pass an array or slice directly that is more efficient.
+
+    A range that is detected as infinite by `std.range.primitives.isInfinite`
+    will be treated as though its length were `n`.
+    +/
+    this(Range)(scope Range range)
+    if (isInputRange!Range && is(Unqual!(ElementType!Range) == UIntType))
```
Comment (9il) [inline source/mir/random/engine/mersenne_twister.d]: this should be three line of code function:
```
import std.algorithm.mutation: copy;
import std.range: take;
this(buffer[0 .. (range.take(n).copy(buffer[]).length].sliced);
```

**Comment by n8sh** [comment]
Including private pure versions pureMalloc and pureFree builds with DMD but not LDC due to this issue:

>Error: Function type does not match previously declared function with the same mangled name: free

So on LDC it looks like the only option is to use impure `free`. This might cause slight annoyance for users because if they call the InputRange constructor from `pure` code then _their_ program will compile with DMD but not on LDC. So if we're not using the GC and not using druntime, then I think the code will have to use impure `malloc`/`free` with all compilers.

File: source/mir/random/engine/mersenne_twister.d (hunk 2 above)
Comment (n8sh) [reply source/mir/random/engine/mersenne_twister.d]: Wouldn't that introduce a dependency on druntime?

Comment (9il) [reply source/mir/random/engine/mersenne_twister.d]: it would not

**Comment by 9il** [comment]
> Including private pureMalloc and pureFree builds with DMD but not LDC due to this issue:
> 
> Error: Function type does not match previously declared function with the same mangled name: free
> So on LDC it looks like the only option is to use impure free. This might cause slight annoyance for users because if they call the InputRange constructor from pure code then their program will compile with DMD but not on LDC. So if we're not using the GC and not using druntime, then I think the code will have to use impure malloc/free with all compilers.

1. You do not need to allocate memory at all because the max used length us know at compile time.
2. Just use this three LOC

**Comment by 9il** [comment]
If so, lets just remove the method for ranges and use existing for array and slice only

**Comment by n8sh** [comment]
>Do we really need to use entire array if it length is larger then N?

If we want it to match the reference implementation.

>If so, lets just remove the method for ranges and use existing for array and slice only

Alright.

**Pull request merged**